Adding live metrics

hectorhdzg opened the pull request:
Fixes #52

hectorhdzg committed:
Adding live metrics

hectorhdzg commented:
Code to expose this to final user is not part of this PR, currently we need to add the AzureMetricsSpanProcessor to the TracerProvider and instantiate the AutoCollection classes, it will be good to have some initial configuration that take care of this

hectorhdzg committed:
Merge branch 'master' into hectorh/liveMetr

lzchen commented on azure_monitor/src/azure_monitor/export/__init__.py:
```diff
@@ -204,3 +209,105 @@ def get_metrics_export_result(result: ExportResult) -> MetricsExportResult:
+
+# pylint: disable=too-many-statements
+# pylint: disable=too-many-branches
+def convert_span_to_envelope(span: Span) -> protocol.Envelope:
```
Why move this here?

hectorhdzg replied:
Need this conversion in LiveMetrics processing, could move to a more appropriate location like utility or something, any suggestion?

lzchen replied:
Can we leave this in trace and have LiveMetrics just import from trace module?

lzchen commented on azure_monitor/src/azure_monitor/export/trace/__init__.py:
```diff
@@ -52,104 +53,6 @@ def export(self, spans: Sequence[Span]) -> SpanExportResult:
-            data.properties["_MS.links"] = json.dumps(links)
-        # TODO: tracestate, tags
+        envelope = convert_span_to_envelope(span)
+        envelope.ikey = self.options.instrumentation_key
```
How come we can't set the ikey in the `convert` function?

hectorhdzg replied:
ikey is not available everywhere in the SDK, is currently specific to exporters, validation code is already in there, will need to centralize ikey management to be able to add the ikey in some shared function

lzchen commented:
Can you add a summary of all the changes in the description (including moving certain files) and also the reasoning behind those changes?

lzchen reviewed (changes requested):
It would also be good to document somewhere in a consolidated place exactly what standard metrics are autocollected and what live metrics are autocollected.

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/live_metrics/dependency_metrics.py:
```diff
@@ -0,0 +1,109 @@
+            self._span_processor.dependency_duration
+            - dependency_map.get("last_duration", 0)
+        )
+        interval_count = (
```
Is there someway to abstract the logic between livemetrics and autocollection? Or is this not possible because they have separate export intervals?

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/metrics_span_processor.py:
```diff
@@ -0,0 +1,63 @@
+    """
+
+    def __init__(self, exporter=None):
+        self._exporter = exporter
```
Having the exporter be optional implies we should have a default one. Otherwise, we should make it mandatory.

hectorhdzg replied:
Exporter removed from here, this was to have a reference to LiveMetricsExporter to be able to send documents to QuickPulse, updated the code to use the span processor in the exporter isntead

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/metrics_span_processor.py:
```diff
@@ -0,0 +1,63 @@
+    def on_end(self, span: Span) -> None:
+        try:
+            if span.kind == SpanKind.SERVER:
+                self.request_count = self.request_count + 1
```
Couple of things to point on with spanprocessor on_end as the indicator to request/dependency happening. 

1.First of all, the counts/durations would be inconsistent depending on how long it takes the span to finish. 

2. Secondly, this requires the user to be instrumented with the required integrations for this to work (spans have to be generated). This might actually be a desired design to have but currently other SDK (Opencensus) does not behave like this. Might be good to discuss.

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/__init__.py:
```diff
@@ -30,7 +35,14 @@ class AutoCollection:
+        self,
+        meter: Meter,
+        labels: Dict[str, str],
+        span_processor: AzureMetricsSpanProcessor,
```
So we expect a user to pass in an exporter into the spanprocessor when instantiating? Seems should be the otherway around, spanprocessor should be configured as part of the exporter.

hectorhdzg replied:
SpanProcessor doesn't have any reference to an exporter anymore

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/live_metrics/dependency_metrics.py:
```diff
@@ -0,0 +1,109 @@
+            value_type=int,
+        )
+
+    def _track_dependency_duration(self, observer) -> None:
```
Missing typing for a lot of these headers?

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/metrics_span_processor.py:
```diff
@@ -0,0 +1,63 @@
+    and failed dependencies/requests.
+    """
+
+    def __init__(self, exporter=None):
```
Typing? What exporters are we expecting? Can a user pass in a LogExporter?

lzchen commented on azure_monitor/src/azure_monitor/sdk/auto_collection/metrics_span_processor.py:
```diff
@@ -0,0 +1,63 @@
+                if not span.status.is_ok:
+                    self.failed_request_count = self.failed_request_count + 1
+                    if self._exporter is not None:
+                        self._exporter.add_document(
```
I think only the livemetricsexporter has `add_document`. Don't think the AzureMonitorMetricsExporter has this for standard metrics.

hectorhdzg committed:
Fixing lint issues

hectorhdzg committed:
Merge branch 'hectorh/liveMetr' of https://github.com/hectorhdzg/opentelemetry-azure-monitor-python into hectorh/liveMetr

hectorhdzg committed:
Using single classes for live/standard metrics to avoid duplication
SpanProcessor used in LiveMetricExporter instead of other way around
Updating example

hectorhdzg committed:
Adding tests

hectorhdzg committed:
Lint

hectorhdzg committed:
Fixing issue with rates as int
No need to differentiate between live and standard metrics in reuqests and dependencies
Add logic to stop collecting documents when live metrics are not subscribed
Fix issues with LiveMetric document creation

hectorhdzg committed:
Fixing issue with documents parsing

lzchen reviewed (approved):
LGTM
